torchvision/prototype/models/depth/stereo/raft_stereo.py

```diff
@@ -0,0 +1,696 @@
+        super(raft.MaskPredictor, self).__init__()
+        self.convrelu = Conv2dNormActivation(in_channels, hidden_size, norm_layer=None, kernel_size=3)
+        self.conv = nn.Conv2d(hidden_size, out_channels, kernel_size=1, padding=0)
+        self.multiplier = multiplier
+
+
+class CorrPyramid1d(nn.Module):
+    """Row-wise correlation pyramid.
+
+    Create a row-wise correlation pyramid with ``num_levels`` level from the outputs of the feature encoder,
+    this correlation pyramid will later be used as index to create correlation features using CorrBlock1d.
+    """
+
+    def __init__(self, num_levels: int = 4):
+        super().__init__()
+        self.num_levels = num_levels
+
+    def forward(self, fmap1: Tensor, fmap2: Tensor) -> List[Tensor]:
+        """Build the correlation pyramid from two feature maps.
+
+        The correlation volume is first computed as the dot product of each pair (pixel_in_fmap1, pixel_in_fmap2) on the same row.
+        The last 2 dimensions of the correlation volume are then pooled num_levels times at different resolutions
+        to build the correlation pyramid.
+        """
+
+        if fmap1.shape != fmap2.shape:
+            raise ValueError(
+                f"Input feature maps should have the same shape, instead got {fmap1.shape} (fmap1.shape) != {fmap2.shape} (fmap2.shape)"
+            )
+
+        # corr_volume = self._compute_corr_volume(fmap1, fmap2)
+        batch_size, num_channels, h, w = fmap1.shape
+        fmap1 = fmap1.view(batch_size, num_channels, h, w)
+        fmap2 = fmap2.view(batch_size, num_channels, h, w)
+
+        corr = torch.einsum("aijk,aijh->ajkh", fmap1, fmap2)
+        corr = corr.view(batch_size, h, w, 1, w)
+        corr_volume = corr / torch.sqrt(torch.tensor(num_channels))
+
+        corr_volume = corr_volume.reshape(batch_size * h * w, 1, 1, w)
+        corr_pyramid = [corr_volume]
+        for _ in range(self.num_levels - 1):
+            corr_volume = F.avg_pool2d(corr_volume, kernel_size=(1, 2), stride=(1, 2))
+            corr_pyramid.append(corr_volume)
+
+        return corr_pyramid
+
+
+class CorrBlock1d(nn.Module):
+    """The row-wise correlation block.
+
+    Use indexes from correlation pyramid to create correlation features.
+    The "indexing" of a given centroid pixel x' is done by concatenating its surrounding row neighbours
+    within radius
+    """
+
+    def __init__(self, *, num_levels: int = 4, radius: int = 4):
+        super().__init__()
+        self.radius = radius
+        self.out_channels = num_levels * (2 * radius + 1)
+
+    def forward(self, centroids_coords: Tensor, corr_pyramid: List[Tensor]) -> Tensor:
+        """Return correlation features by indexing from the pyramid."""
+        neighborhood_side_len = 2 * self.radius + 1  # see note in __init__ about out_channels
+        di = torch.linspace(-self.radius, self.radius, neighborhood_side_len)
```

Comment:
Similarly, let's put a device here and in all other places where you create new tensors.